app/fileUploadParsers/AnnualAllowanceParser.scala

```diff
@@ -17,26 +17,101 @@
+                                       config: FrontendAppConfig
+                                     ) extends Parser with Constraints {
+  //scalastyle:off magic.number
   override protected val totalFields: Int = 7
 
-  override protected def validateFields(ua: UserAnswers, index: Int, chargeFields: Array[String]): Either[ParserValidationErrors, UserAnswers] = {
-    val m = Map(
-      "firstName" -> firstNameField(chargeFields),
-      "lastName" -> lastNameField(chargeFields),
-      "nino" -> ninoField(chargeFields)
-    )
-    val form = memberDetailsFormProvider.apply()
-    form.bind(m).fold(
-      formWithErrors => Left(ParserValidationErrors(index, formWithErrors.errors.map(_.message))),
-      value => Right(ua.setOrException(MemberDetailsPage(index), value))
+  override protected def validHeader: String = config.validAnnualAllowanceHeader
+
+  private final object ChargeDetailsFieldNames {
+    val chargeAmount: String = "chargeAmount"
+    val dateNoticeReceivedDay: String = "dateNoticeReceived.day"
+    val dateNoticeReceivedMonth: String = "dateNoticeReceived.month"
+    val dateNoticeReceivedYear: String = "dateNoticeReceived.year"
+    val dateNoticeReceived: String = "dateNoticeReceived"
+    val isPaymentMandatory = "isPaymentMandatory"
+  }
+
+  private final object TaxYearErrorKeys {
+    val requiredKey = "annualAllowanceYear.fileUpload.error.required"
+    val invalidKey = "annualAllowanceYear.fileUpload.error.invalid"
+    val minKey = "annualAllowanceYear.fileUpload.error.past"
+    val maxKey = "annualAllowanceYear.fileUpload.error.future"
+  }
+
+  private def chargeDetailsValidation(startDate: LocalDate, index: Int, chargeFields: Array[String]): Either[Seq[ParserValidationError], ChargeEDetails] = {
+    val taxYearsErrors = validateTaxYear(startDate, index, chargeFields(3))
+    splitDayMonthYear(chargeFields(5)) match {
+      case Tuple3(day, month, year) =>
+        val fields = Seq(
+          Field(ChargeDetailsFieldNames.chargeAmount, chargeFields(4), ChargeDetailsFieldNames.chargeAmount, 4),
+          Field(ChargeDetailsFieldNames.dateNoticeReceivedDay, day, ChargeDetailsFieldNames.dateNoticeReceived, 5),
+          Field(ChargeDetailsFieldNames.dateNoticeReceivedMonth, month, ChargeDetailsFieldNames.dateNoticeReceived, 5),
+          Field(ChargeDetailsFieldNames.dateNoticeReceivedYear, year, ChargeDetailsFieldNames.dateNoticeReceived, 5),
+          Field(ChargeDetailsFieldNames.isPaymentMandatory, stringToBoolean(chargeFields(6)),
+            ChargeDetailsFieldNames.isPaymentMandatory, 6)
+
+        )
+        val chargeDetailsForm: Form[ChargeEDetails] = chargeDetailsFormProvider(
+          minimumChargeValueAllowed = minChargeValueAllowed,
+          minimumDate = config.earliestDateOfNotice
+        )
+        chargeDetailsForm.bind(
+          Field.seqToMap(fields)
+        ).fold(
+          formWithErrors => Left(errorsFromForm(formWithErrors, fields, index) ++ taxYearsErrors),
```

Comment:
suggest moving the code from 81 to 86 to a private method.

Reply:
I don't really understand this. It's only 4 lines - how would that help code readability?